From baf7d817f44b0357777aefc77d04350a4cbbf8b3 Mon Sep 17 00:00:00 2001
From: Wei Li <liwei391@huawei.com>
Date: Thu, 9 Feb 2023 17:03:58 +0800
Subject: [PATCH 888/905] [Huawei] locking/legacy-percpu-rwsem: Add control
 config

Offering: HULK
hulk inclusion
category: feature
bugzilla: NA

--------------------------------

Add config CONFIG_LEGACY_PERCPU_RWSEM to control legacy-percpu-rwsem,
fallback to percpu-rwsem if CONFIG_LEGACY_PERCPU_RWSEM is not set.

Signed-off-by: Wei Li <liwei391@huawei.com>
---
 include/linux/legacy-percpu-rwsem.h | 16 ++++++++++++++++
 kernel/Kconfig.locks                |  4 ++++
 kernel/locking/Makefile             |  3 ++-
 3 files changed, 22 insertions(+), 1 deletion(-)

diff --git a/include/linux/legacy-percpu-rwsem.h b/include/linux/legacy-percpu-rwsem.h
index 2b011fb9a44e..abb33ebf6226 100644
--- a/include/linux/legacy-percpu-rwsem.h
+++ b/include/linux/legacy-percpu-rwsem.h
@@ -2,6 +2,21 @@
 #ifndef _LINUX_LEGACY_PERCPU_RWSEM_H
 #define _LINUX_LEGACY_PERCPU_RWSEM_H
 
+#ifndef CONFIG_LEGACY_PERCPU_RWSEM
+#include <linux/percpu-rwsem.h>
+
+#define DEFINE_LEGACY_PERCPU_RWSEM(name)	DEFINE_PERCPU_RWSEM(name)
+#define DEFINE_STATIC_LEGACY_PERCPU_RWSEM(name)	DEFINE_STATIC_PERCPU_RWSEM(name)
+
+#define legacy_percpu_down_read(sem)	percpu_down_read(sem)
+#define legacy_percpu_up_read(sem)	percpu_up_read(sem)
+
+#define legacy_percpu_down_write(sem)	percpu_down_write(sem)
+#define legacy_percpu_up_write(sem)	percpu_up_write(sem)
+
+#define legacy_percpu_rw_semaphore	percpu_rw_semaphore
+
+#else
 #include <linux/atomic.h>
 #include <linux/rwsem.h>
 #include <linux/percpu.h>
@@ -147,3 +162,4 @@ static inline void legacy_percpu_rwsem_acquire(struct legacy_percpu_rw_semaphore
 }
 
 #endif
+#endif
diff --git a/kernel/Kconfig.locks b/kernel/Kconfig.locks
index 3de8fd11873b..33fb9b7ab44d 100644
--- a/kernel/Kconfig.locks
+++ b/kernel/Kconfig.locks
@@ -259,3 +259,7 @@ config ARCH_HAS_MMIOWB
 config MMIOWB
 	def_bool y if ARCH_HAS_MMIOWB
 	depends on SMP
+
+config LEGACY_PERCPU_RWSEM
+	bool "The legacy percpu rw semaphores"
+	default n
diff --git a/kernel/locking/Makefile b/kernel/locking/Makefile
index efac4b446306..ad53e85bda4c 100644
--- a/kernel/locking/Makefile
+++ b/kernel/locking/Makefile
@@ -3,7 +3,7 @@
 # and is generally not a function of system call inputs.
 KCOV_INSTRUMENT		:= n
 
-obj-y += mutex.o semaphore.o rwsem.o percpu-rwsem.o legacy-percpu-rwsem.o
+obj-y += mutex.o semaphore.o rwsem.o percpu-rwsem.o
 
 # Avoid recursion lockdep -> KCSAN -> ... -> lockdep.
 KCSAN_SANITIZE_lockdep.o := n
@@ -32,3 +32,4 @@ obj-$(CONFIG_QUEUED_RWLOCKS) += qrwlock.o
 obj-$(CONFIG_LOCK_TORTURE_TEST) += locktorture.o
 obj-$(CONFIG_WW_MUTEX_SELFTEST) += test-ww_mutex.o
 obj-$(CONFIG_LOCK_EVENT_COUNTS) += lock_events.o
+obj-$(CONFIG_LEGACY_PERCPU_RWSEM) += legacy-percpu-rwsem.o
-- 
2.17.1

